shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/Spark300Shims.scala

```diff
@@ -436,4 +445,68 @@ class Spark300Shims extends SparkShims {
       explicitMetadata: Option[Metadata]): Alias = {
     Alias(child, name)(exprId, qualifier, explicitMetadata)
   }
+
+  override def getLocationWithAlluxioReplace(
+      conf: RapidsConf,
+      relation: HadoopFsRelation,
+      partitionFilters: Seq[Expression],
+      dataFilters: Seq[Expression]): FileIndex = {
+
+    val alluxioFilesReplace: Option[Seq[String]] = conf.getAlluxioFilesReplace
+
+    if (alluxioFilesReplace.isDefined) {
+      // alluxioFilesReplace: Seq("key->value", "key1->value1")
+      val replaceMap = new mutable.HashMap[String, String]()
+      alluxioFilesReplace.get.foreach(rule => {
+        val ruleSplit = rule.split("->")
+        if (ruleSplit.size == 2) {
+          replaceMap += ruleSplit(0).trim -> ruleSplit(1).trim
+        } else {
+          logWarning("Wrong set for spark.rapids.alluxio.pathsToReplace " + rule)
+        }
+      })
+
+      if (replaceMap.size > 0) {
+        val listFiles: Seq[PartitionDirectory] = relation.location.listFiles(
+          partitionFilters, dataFilters)
+
+        val replaceFunc = (f: Path) => {
+          val matchedSet = replaceMap.keySet.filter(reg => f.toString.startsWith(reg))
+          if (matchedSet.size > 0) {
+            new Path(
+              f.toString.replaceFirst(matchedSet.head, replaceMap(matchedSet.head)))
+          } else {
+            f
+          }
+        }
+
+        val inputFiles: Seq[Path] = listFiles.flatMap(partitionDir => {
+          partitionDir.files.map(f => replaceFunc(f.getPath))
+        }).toSet.toSeq
+
+        // get the leaf dir of inputFiles
+        val leafDirs = inputFiles.map(_.getParent).toSet.toSeq
+
+        val finalRootPaths = relation.location.rootPaths.map(replaceFunc)
+
+        val partitionSpec = RapidsPartitioningUtils.inferPartitioning(
+          relation.sparkSession,
+          leafDirs,
+          finalRootPaths.toSet,
+          relation.options,
+          Option(relation.dataSchema))
+
+        new InMemoryFileIndex(
```

Comment:
so I believe this will submit another job to spark to do the listing when its over 32 files, can you confirm? How long does that job take?  Not sure there is anything we can do or want to do about it, but just wondering

Reply:
so can you confirm it did run a separate job and that took < 1s?

Reply:
Yes, confirmed, I added some log to metric the function replaceWithAlluxioPathIfNeeded for the query based on 4152 files and the first run took longer (1.876, 6.26, 1.222, 1.171)s, but the 2nd run only took about 1s (0.049, 1.162, 0.845, 0.748)s.

Reply:
I'm assuming those numbers are the time of the extra job. I would expect the string replacement to take the same amount of time each run.  But I'm guessing the difference is in the actual listing.